cis194/week5/zhansongl/Calc.hs

```diff
@@ -0,0 +1,95 @@
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE FlexibleInstances #-}
+module Calc where
+
+import ExprT
+import Parser
+import Control.Applicative (liftA, liftA2)
+import qualified StackVM as S
+import qualified Data.Map as M
+
+eval :: ExprT -> Integer
+eval (Lit n) = n
+eval (Add e1 e2) = eval e1 + eval e2
+eval (Mul e1 e2) = eval e1 * eval e2
+
+evalStr :: String -> Maybe Integer
+evalStr = liftA eval . parseExp Lit Add Mul
+
+class Expr a where
+  lit :: Integer -> a
+  add :: a -> a -> a
+  mul :: a -> a -> a
+
+instance Expr ExprT where
+  lit = Lit
+  add = Add
+  mul = Mul
+
+reify :: ExprT -> ExprT
+reify = id
+
+instance Expr Integer where
+  lit = id
+  add = (+)
+  mul = (*)
+
+instance Expr Bool where
+  lit = (>0)
+  add = (||)
+  mul = (&&)
+
+newtype MinMax = MinMax Integer deriving (Eq, Show)
+newtype Mod7   = Mod7 Integer deriving (Eq, Show)
+
+instance Expr MinMax where
+  lit = MinMax
+  add (MinMax x) (MinMax y) = MinMax $ max x y
+  mul (MinMax x) (MinMax y) = MinMax $ min x y
+
+instance Expr Mod7 where
+  lit = Mod7 . (`mod` 7)
+  add (Mod7 x) (Mod7 y) = Mod7 $ (x + y) `mod` 7
+  mul (Mod7 x) (Mod7 y) = Mod7 $ (x * y) `mod` 7
+
+testExp :: Expr a => String -> Maybe a
+testExp = parseExp lit add mul
+
+instance Expr S.Program where
+  lit n = [S.PushI n]
+  add p1 p2 = p1 ++ p2 ++ [S.Add]
```

Comment:
What if `p1` or `p2` is `[BVal True]`?

Reply:
hmm, I'm not sure actually, is this supposed to work with BVal?

Reply:
Oh I re-read the question again, looks like we should only worry about `IVal`...

The module `StackVM` is a bit misleading I think... Makes me overthink too much.

Reply:
> For any arithmetic expression exp :: Expr a => a it should be the case that
`stackVM exp == Right [IVal exp]`

I think this sentence is wrong `stackVM exp == Right [IVal exp]`

It should be `stackVM exp == Right [IVal val]` @shineli1984 Am I correct?

Reply:
Honestly, I can understand the question, but wasn't sure what I was supposed to do with this. I think this solution is correct though :)